modules/aruco/src/dictionary.cpp

```diff
@@ -0,0 +1,473 @@
+                currentHamming += hammingWeightLUT[xorRes];
+            }
+
+            if(currentHamming < currentMinDistance) {
+                currentMinDistance = currentHamming;
+                currentRotation = r;
+            }
+        }
+
+        // if maxCorrection is fullfilled, return this one
+        if(currentMinDistance <= maxCorrectionRecalculed) {
+            idx = m;
+            rotation = currentRotation;
+            break;
+        }
+    }
+
+    if(idx != -1)
+        return true;
+    else
+        return false;
+}
+
+
+/**
+  */
+int Dictionary::getDistanceToId(InputArray bits, int id, bool allRotations) const {
+
+    CV_Assert(id >= 0 && id < bytesList.rows);
+
+    unsigned int nRotations = 4;
+    if(!allRotations) nRotations = 1;
+
+    Mat candidateBytes = getByteListFromBits(bits.getMat());
+    int currentMinDistance = int(bits.total() * bits.total());
+    for(unsigned int r = 0; r < nRotations; r++) {
+        int currentHamming = 0;
+        for(unsigned int b = 0; b < candidateBytes.total(); b++) {
+            unsigned char xorRes =
+                bytesList.ptr< Vec4b >(id)[b][r] ^ candidateBytes.ptr< Vec4b >(0)[b][0];
+            currentHamming += hammingWeightLUT[xorRes];
+        }
+
+        if(currentHamming < currentMinDistance) {
+            currentMinDistance = currentHamming;
+        }
+    }
+    return currentMinDistance;
+}
+
+
+
+/**
+ * @brief Draw a canonical marker image
+ */
+void Dictionary::drawMarker(int id, int sidePixels, OutputArray _img, int borderBits) const {
+
+    CV_Assert(sidePixels > markerSize);
+    CV_Assert(id < bytesList.rows);
+    CV_Assert(borderBits > 0);
+
+    _img.create(sidePixels, sidePixels, CV_8UC1);
+
+    // create small marker with 1 pixel per bin
+    Mat tinyMarker(markerSize + 2 * borderBits, markerSize + 2 * borderBits, CV_8UC1,
+                   Scalar::all(0));
+    Mat innerRegion = tinyMarker.rowRange(borderBits, tinyMarker.rows - borderBits)
+                          .colRange(borderBits, tinyMarker.cols - borderBits);
+    // put inner bits
+    Mat bits = 255 * getBitsFromByteList(bytesList.rowRange(id, id + 1), markerSize);
+    CV_Assert(innerRegion.total() == bits.total());
+    bits.copyTo(innerRegion);
+
+    // resize tiny marker to output size
+    cv::resize(tinyMarker, _img.getMat(), _img.getMat().size(), 0, 0, INTER_NEAREST);
+}
+
+
+
+
+/**
+  * @brief Transform matrix of bits to list of bytes in the 4 rotations
+  */
+Mat Dictionary::getByteListFromBits(const Mat &bits) {
+
+    int nbytes = (bits.cols * bits.rows) / 8;
+    if((bits.cols * bits.rows) % 8 != 0) nbytes++;
+
+    Mat candidateByteList(1, nbytes, CV_8UC4, Scalar::all(0));
+    unsigned char currentBit = 0;
+    int currentByte = 0;
+    for(int row = 0; row < bits.rows; row++) {
+        for(int col = 0; col < bits.cols; col++) {
+            // circular shift
+            candidateByteList.ptr< Vec4b >(0)[currentByte][0] =
```

Comment:
Minor optimization: You could create a temp variable/reference to ```candidateByteList.ptr< Vec4b >(0)[currentByte]``` and manipulate it inside the loop. Here you are accessing the same thing again. The processor caching would definitely help you in the current state. But this way it s easy to read as well.